### ISSUE TITLE
Performance drops massively after a couple of minutes (GPU down to 0%)

### ISSUE THREAD

**Issue by pappitti** [ml-explore/mlx#868]
**Describe the bug**
After a precise number of operations on very large matrices, GPU support seems to break and I suspect that CPU is significantly throttled too.  
Trying to reproduce, I experienced the same issue (not exactly after the same number of operations but very close) with different hardware and different OS. I ruled out issues related to recursion as I reproduced the issue with iterations only and an equivalent script using numpy (and recursions) works in colab. I also ruled out python/environment issues as an equivalent script with pytorch (with metal support) also runs flawlessly on my two devices in the same environments as the MLX ones.  
I think that it may be MLX-related.

https://github.com/ml-explore/mlx/assets/90358606/46cb2f98-6fcf-43ac-bb94-87cef9bfa51f

**To Reproduce**

The code below was created for a small educational project using sudokus to illustrate how matrix operations can be leveraged. The goal is to produce all possible sudokus. It is pretty basic in the sense that it generates all possible rows (permutations of digits) and then uses recursion to stack them together so as to respect the sudoku constraints along the first dimension and in the sub-squares.  

```  
import mlx.core as mx
import time
from itertools import permutations

print(mx.default_device())

Y=3 # size of the sub-squares (3 for a real sudoku)
X=Y**2 # size of the sudoku (9 for a real sudoku)

start_time = time.time()

# generate vectors with 0s and 1s corresponding to each digit in your grid
base_vectors=[]

for i in range(X):
    base_vector=mx.zeros(X) 
    base_vector[i]=1
    base_vectors.append(base_vector)

# generate possible rows (i.e. permutations of digits) - there are 24 for X=4 and 362,880 for X=9
all_permutations= permutations(base_vectors, X)
sudoku_rows=[mx.stack(i) for i in all_permutations]

print("1D : ", len(sudoku_rows))
print("1D --- %s seconds ---" % (time.time() - start_time))

# generate grids
sudokus=[]

class Sudoku:
    def __init__(self,seed=mx.array([]),possible_values=sudoku_rows):
        self.values=seed
        self.possible_values=possible_values

    def populate(self):

        if self.values.size:
            mat_sum=mx.sum(self.values, axis=0, stream=mx.cpu)
            current_row=self.values.shape[0]
        else:
            mat_sum=mx.zeros([X,X],stream=mx.cpu)
            current_row=0

        vertical_filter_vector=mx.subtract(mx.ones([X,X],stream=mx.cpu),mat_sum, stream=mx.cpu)

        possible_values_matrice=mx.stack(self.possible_values, stream=mx.gpu)

        # here apply element-wise multiplication on possible values so as to avoid slicing the seed for matmul, and we only sum the resulting matrice at the end
        element_wise_multiplication=mx.multiply(possible_values_matrice,vertical_filter_vector, stream=mx.gpu)

        collapse_z=mx.sum(element_wise_multiplication,axis=2, stream=mx.gpu)
        collapse_y=mx.sum(collapse_z,axis=1, stream=mx.gpu)

        vertical_possible_values=[self.possible_values[i] for i,v in enumerate(collapse_y.tolist()) if v==X**1]
        
        # case where we have a sub-square
        if current_row%Y!=0:
            sub_square=current_row//Y
            sub_row=current_row%Y

            # only keep the relevant rows
            sliced_array=self.values[sub_square*Y:sub_square*Y+sub_row,:,:]
            sub_mat_sum=mx.sum(sliced_array, axis=0, stream=mx.cpu)
           
            # break the rows down into columns
            sub_mat=mx.reshape(sub_mat_sum,(Y,Y,X), stream=mx.cpu)
            
            sub_mat_collapsed=mx.sum(sub_mat,axis=1, stream=mx.cpu)

            # rebuild an array of shape [X,X] by duplicating vectors along the first axis
            repeated_sub_mat=mx.repeat(sub_mat_collapsed,Y,axis=0, stream=mx.cpu)
           
            square_filter_vector=mx.subtract(mx.ones([X,X], stream=mx.cpu),repeated_sub_mat, stream=mx.cpu)
           
            possible_values_matrice=mx.stack(vertical_possible_values, stream=mx.gpu)

            element_wise_multiplication=mx.multiply(possible_values_matrice,square_filter_vector, stream=mx.gpu)
            
            collapse_z=mx.sum(element_wise_multiplication,axis=2, stream=mx.gpu)
            collapse_y=mx.sum(collapse_z,axis=1, stream=mx.gpu)

            filtered_possible_values=[vertical_possible_values[i] for i,v in enumerate(collapse_y.tolist()) if v==X**1]
        else:
            filtered_possible_values=vertical_possible_values

        # add new row and start population process for next one
        for i in filtered_possible_values:
            new_row=mx.reshape(i,shape=(1,X,X), stream=mx.cpu)
            populated_rows=mx.concatenate((self.values,new_row), stream=mx.cpu) if self.values.size else new_row
            if current_row==X-1:
                sudokus.append(populated_rows)
                print(populated_rows.shape, len(sudokus))
                return
            else:
                next_position=Sudoku(populated_rows, vertical_possible_values)
                next_position.populate()

sudoku_start=Sudoku()
sudoku_start.populate()

print("2D: ", len(sudokus))
print("2D --- %s seconds ---" % (time.time() - start_time))
```

**Expected behavior**
The script above records an entry each time a full grid is generated. When setting Y=3 (for 9x9 sudoku grids) the issue arises after ~133900 entries. The expected behavior is that it continues using 100% of GPU capacity until it is finished. Testing with Y=2 (4x4 grids) allows to test the full cycle up to 2D. 

**Devices:**
 - OS Version: MacOS 13.5 / M2 Ultra 192 GB and MacOS 14.1.2 / M2 Pro 16GB
 - Version : 0.6.0 and 0.7.0
  
**Additional Information**
As MLX differs slightly from numpy or pytorch (e.g. masking/filtering where I had to be creative), I could not be sure that my implementation is correct. For reference, I paste below the working code with pytorch. You will likely run into memory error with pytorch ( I had to change approach but it is unrelated with the GPU bug)  

```
import torch
import numpy as np
from itertools import permutations
import time

if torch.backends.mps.is_available():
    mps_device = torch.device("mps")
    x = torch.ones(1, device=mps_device)
    print (x)
else:
    print ("MPS device not found.")

start_time = time.time()

Y=2 # size of the sub-squares (3 for a real sudoku)
X=Y**2 # size of the sudoku (9 for a real sudoku)

base_vectors=[]

for i in range(X):
    base_vector=torch.zeros(X, dtype=torch.int8)
    base_vector[i]=1
    base_vectors.append(base_vector)

all_permutations = permutations(base_vectors, X)
sudoku_rows = [torch.stack(permutation) for permutation in all_permutations]

sudokus=[]

class Sudoku:
    def __init__(self,seed=torch.empty((0,)),possible_values=torch.stack(sudoku_rows)):
        self.values=seed
        self.possible_values=possible_values

    def populate(self):
        if self.values.nelement():
            mat_sum = torch.sum(self.values, dim=0)
            current_row = self.values.shape[0]
        else:
            mat_sum = torch.zeros((X, X), dtype=torch.int8, device=mps_device)
            current_row = 0

        vertical_filter_vector = torch.sub(torch.ones((X, X), dtype=torch.int8, device=mps_device), mat_sum)

        possible_values_matrix = self.possible_values.to(mps_device)

        # here apply element-wise multiplication on possible values so as to avoid slicing the seed for matmul, and we only sum the resulting matrice at the end
        element_wise_multiplication = torch.mul(possible_values_matrix, vertical_filter_vector)

        sum_to_vector=torch.sum(element_wise_multiplication,(2,1))

        mask = (sum_to_vector== X**1)
        
        indices=torch.nonzero(mask, as_tuple=False).flatten()

        vertical_possible_values=torch.index_select(possible_values_matrix, 0, indices)
        
        # case where we have a sub-square
        if current_row % Y != 0:
            sub_square = current_row // Y
            sub_row = current_row % Y
            sliced_array = self.values[sub_square*Y:sub_square*Y+sub_row, :, :]
            sub_mat_sum = torch.sum(sliced_array, dim=0)

            sub_mat = torch.reshape(sub_mat_sum, (Y, Y, X))

            sub_mat_collapsed = torch.sum(sub_mat, dim=1)

            repeated_sub_mat = torch.repeat_interleave(sub_mat_collapsed, Y, dim=0)

            square_filter_vector = torch.sub(torch.ones((X, X), dtype=torch.int8, device=mps_device), repeated_sub_mat)

            possible_values_matrix = vertical_possible_values

            element_wise_multiplication = torch.mul(possible_values_matrix, square_filter_vector)

            sum_to_vector=torch.sum(element_wise_multiplication,(2,1))

            mask = (sum_to_vector== X**1)
            
            indices=torch.nonzero(mask, as_tuple=False).flatten()

            filtered_possible_values=torch.index_select(vertical_possible_values, 0, indices)
        else:
            filtered_possible_values=vertical_possible_values
    
        for i in filtered_possible_values:
            new_row = torch.reshape(i, (1, X, X))
            populated_rows = torch.cat((self.values, new_row), dim=0) if self.values.nelement() else new_row
            if current_row==X-1:
                sudokus.append(populated_rows)
                print(populated_rows.shape, len(sudokus))
                return
            else:
                next_position=Sudoku(populated_rows, vertical_possible_values)
                next_position.populate()

sudoku_start=Sudoku()
sudoku_start.populate()

all_sudokus=torch.stack(sudokus).cpu().numpy()
print("2D : ",all_sudokus.shape)

np.save(f"sudokus_dim{X}.npy",all_sudokus)
print("2D --- %s seconds ---" % (time.time() - start_time))
```

**Comment by awni** [comment]
So the reason this code hangs is you are running into a limit on the maximum number of MTL buffers one can allocate with a single process.

For example, this line is making a really large list (350k):

```
sudoku_rows = [torch.stack(permutation) for permutation in all_permutations]
```

As is this line:

```
                sudokus.append(populated_rows)
```


Ideally I would not hold onto all of those arrays. Another workaround would  be to use Python lists for some of those instead of `mx.array`. I don't think it's likely we will fix that since its a system limit, but it may be possible to track and crash with a more helpful error message.

**Comment by pappitti** [comment]
Thanks for looking into it. I did not know about the limit on MTL buffers. There was never an error message or I did not know where to look that's why it was so puzzling. All the more that it is not an issue with other frameworks. I fully understand that this is not am issue that is worth spending time and resources on. It is a math problem, not an ML application.

And to your point on de length of the list, I knew before I even started (the problem looks simple but it is linked to many other problems that have never been solved mathematically, only through brute force). I thought my Mac Studio could handle it and I wanted to give MLX a spin, but the algorithm above is absolutely not the solution (even with other frameworks).  I have moved on.

Finally, to your question on streaming on CPU vs GPU, I thought that I followed the doc (there was an example showing some benefits of running small operations on CPU and large ones on GPU) but I may have misunderstood.

**Comment by awni** [comment]
>  I did not know about the limit on MTL buffers. There was never an error message or I did not know where to look that's why it was so puzzling

I didn't know about the limit either 😓 , and yea we should really try to report an error for that. I will look into it.

> Finally, to your question on streaming on CPU vs GPU, I thought that I followed the doc (there was an example showing some benefits of running small operations on CPU and large ones on GPU) but I may have misunderstood.

You did not misunderstand. However, its quite tricky to get the CPU and GPU to co-operate efficiently so usually I would start with a single stream and then try to offload work after profiling a bit.
